packages/apputils/src/widgettracker.ts

```diff
@@ -0,0 +1,366 @@
+  /**
+   * A signal emitted when a widget is added.
+   */
+  readonly widgetAdded: ISignal<this, T>;
+
+  /**
+   * The current widget is the most recently focused or added widget.
+   *
+   * #### Notes
+   * It is the most recently focused widget, or the most recently added
+   * widget if no widget has taken focus.
+   */
+  readonly currentWidget: T | null;
+
+  /**
+   * A signal emitted when the current instance changes.
+   *
+   * #### Notes
+   * If the last instance being tracked is disposed, `null` will be emitted.
+   */
+  readonly currentChanged: ISignal<this, T | null>;
+
+  /**
+   * The number of instances held by the tracker.
+   */
+  readonly size: number;
+
+  /**
+   * A promise that is resolved when the instance tracker has been
+   * restored from a serialized state.
+   *
+   * #### Notes
+   * Most client code will not need to use this, since they can wait
+   * for the whole application to restore. However, if an extension
+   * wants to perform actions during the application restoration, but
+   * after the restoration of another instance tracker, they can use
+   * this promise.
+   */
+  readonly restored: Promise<void>;
+
+  /**
+   * A signal emitted when a widget is updated.
+   */
+  readonly widgetUpdated: ISignal<this, T>;
+
+  /**
+   * Find the first instance in the tracker that satisfies a filter function.
+   *
+   * @param - fn The filter function to call on each instance.
+   *
+   * #### Notes
+   * If nothing is found, the value returned is `undefined`.
+   */
+  find(fn: (obj: T) => boolean): T | undefined;
+
+  /**
+   * Iterate through each instance in the tracker.
+   *
+   * @param fn - The function to call on each instance.
+   */
+  forEach(fn: (obj: T) => void): void;
+
+  /**
+   * Filter the instances in the tracker based on a predicate.
+   *
+   * @param fn - The function by which to filter.
+   */
+  filter(fn: (obj: T) => boolean): T[];
+
+  /**
+   * Check if this tracker has the specified instance.
+   *
+   * @param obj - The object whose existence is being checked.
+   */
+  has(obj: Widget): boolean;
+
+  /**
+   * Inject an instance into the instance tracker without the tracker handling
+   * its restoration lifecycle.
+   *
+   * @param obj - The instance to inject into the tracker.
+   */
+  inject(obj: T): void;
+}
+
+/**
+ * A class that keeps track of widget instances on an Application shell.
+ *
+ * @typeparam T - The type of widget being tracked. Defaults to `Widget`.
+ *
+ * #### Notes
+ * The API surface area of this concrete implementation is substantially larger
+ * than the widget tracker interface it implements. The interface is intended
+ * for export by JupyterLab plugins that create widgets and have clients who may
+ * wish to keep track of newly created widgets. This class, however, can be used
+ * internally by plugins to restore state as well.
+ */
+export class WidgetTracker<T extends Widget = Widget>
```

Comment:
Likewise, I was a bit surprised to see this not extending from `InstanceTracker<T>`